### PR TITLE
Spring pet store test case

### LINKED ISSUE #217: Allow spring-web extension REST endpoints to be published in openapi
Reported in quarkusio/quarkus#4693

Quarkus uses SmallRye OpenAPI to render API endpoints. As reported, if you create a Spring Web REST endpoint and add openapi, the Spring endpoint does not show up.

The OpenApiAnnotationScanner (or an extension) will need to discover Spring Rest endpoints.

### PR DESCRIPTION
**Pull request by ebullient**
This adds a test case based on the generated server classes for the Swagger Pet Store example.

This is not ready for integration yet, as the result is missing all kinds of information (and it does a few things slightly differently than other tools, which isn't a problem, but will require a custom "good" result json for testing once it all works.

Resolves #217

### CODE REVIEW

File: extension-spring/pom.xml [hunk 1]
```diff
@@ -56,7 +56,18 @@
             <version>2.5</version>
             <scope>test</scope>
         </dependency>
-
+        <dependency>
+            <groupId>javax.validation</groupId>
+            <artifactId>validation-api</artifactId>
+            <version>2.0.0.Final</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (MikeEdgar) [inline extension-spring/pom.xml:64]: Please update this to use jakarta.validation:jakarta.validation-api:2.0.2

File: extension-spring/src/main/java/io/smallrye/openapi/spring/SpringAnnotationScanner.java [hunk 2]
```diff
@@ -67,7 +68,7 @@ public boolean isPostMethod(final MethodInfo method) {
 
     @Override
     public boolean isScannerInternalResponse(Type returnType) {
-        return returnType.name().equals(SpringConstants.RESPONSE_ENTITY);
+        return returnType.name().equals(SpringConstants.RESPONSE_ENTITY) && returnType.kind() != Kind.PARAMETERIZED_TYPE;
```
Comment (MikeEdgar) [inline extension-spring/src/main/java/io/smallrye/openapi/spring/SpringAnnotationScanner.java:71]: I think you might want to check for an instance of `HttpEntity` where the type argument is a wildcard, e.g. `HttpEntity<?>`. `ResponseEntity` extends `HttpEntity` so you will cover a few more scenarios.

See `OpenApiDataObjectScanner`'s static init block for an example of how to get the `ClassInfo` of types that cannot be assumed to be in the application's `IndexView` and also how to check if a type inherits from another using `TypeUtil#isA`.

Related to the type of response, you may also look at creating an implementation of `AnnotationScannerExtension` that is used in the Spring scanner with the `resolveAsyncType` overridden. You would want logic similar to that in `SchemaFactory#resolveAsyncType` to pull the type argument from the `HttpEntity<T>` or `ResponseEntity<T>`. The overridden method would then be called from `SchemaFactory#resolveAsyncType`.

File: extension-spring/src/test/java/io/smallrye/openapi/runtime/scanner/SpringAnnotationScannerTest.java [hunk 3]
```diff
@@ -33,4 +33,23 @@ public void testBasicSpringDefinitionScanning() throws IOException, JSONExceptio
...
+        index(indexer, "test/io/smallrye/openapi/runtime/scanner/entities/Order.class");
+        index(indexer, "test/io/smallrye/openapi/runtime/scanner/entities/Pet.class");
+        index(indexer, "test/io/smallrye/openapi/runtime/scanner/entities/Tag.class");
+        index(indexer, "test/io/smallrye/openapi/runtime/scanner/entities/User.class");
+        OpenApiAnnotationScanner scanner = new OpenApiAnnotationScanner(emptyConfig(), indexer.complete());
+
+        OpenAPI result = scanner.scan();
+
+        printToConsole(result);
+        assertJsonEquals("resource.testBasicSpringDefinitionScanning.json", result);
```
Comment (MikeEdgar) [inline extension-spring/src/test/java/io/smallrye/openapi/runtime/scanner/SpringAnnotationScannerTest.java:52]: Is there a different test JSON file that should be used here?

Comment (ebullient) [reply extension-spring/src/test/java/io/smallrye/openapi/runtime/scanner/SpringAnnotationScannerTest.java:52]: I didn't bother to make one yet. ;)
I said this was a draft. What was being produced is still unique to smallrye (in terms of which default strings are used). I will make a json result when what we're producing is close enough that I can define what the correct result is.

File: extension-spring/pom.xml (hunk 1 above)
Comment (ebullient) [reply extension-spring/pom.xml:64]: Sure. I was pulling from other projects.. not an issue.

File: extension-spring/src/main/java/io/smallrye/openapi/spring/SpringAnnotationScanner.java (hunk 2 above)
Comment (ebullient) [reply extension-spring/src/main/java/io/smallrye/openapi/spring/SpringAnnotationScanner.java:71]: I am not sure I understand what that "isScannerInternalResponse" is even supposed to do for Spring. JAX-RS Response is truly useless when it comes to figuring out response types. I'm tempted to always return false, to be honest.

Comment (MikeEdgar) [reply extension-spring/src/main/java/io/smallrye/openapi/spring/SpringAnnotationScanner.java:71]: You're right, but if the parameter is a wildcard, that's no better than the JAX-RS `Response`.

Comment (phillip-kruger) [reply extension-spring/src/main/java/io/smallrye/openapi/spring/SpringAnnotationScanner.java:71]: Yea I think for now return false, if we later figure out there is some class that is like Response then we update it.

**Comment by phillip-kruger** [comment]
Hi @ebullient - should we retry this with the changes I made ? Let me know.

**Comment by phillip-kruger** [comment]
Hi @ebullient - I want to see if this generates a better schema now, but I can not update the branch (here by clicking the button) . Do you want to look at this again ?

**Comment by phillip-kruger** [comment]
No feedback. Closing.

**Pull request closed without merging**